### PR TITLE
[NTUSER] Implement NtUserDisableThreadIme

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Add `IS_IMM_MODE()` macro to `ntuser.h`.
- Add `W32PF_DISABLEIME` flag to `win32.h`.
- Modify `NtUserDisableThreadIme` prototype.
- Implement `NtUserDisableThreadIme` function.

### CODE REVIEW

File: win32ss/user/ntuser/ime.c [hunk 1]
```diff
@@ -80,14 +80,72 @@ NtUserCheckImeHotKey(
...
+
+    if (!IS_IMM_MODE())
+    {
+        EngSetLastError(ERROR_CALL_NOT_IMPLEMENTED);
+        goto Quit;
+    }
+
+    ptiCurrent = GetW32ThreadInfo();
+
+    if (dwThreadID == (DWORD)-1)
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:102]: These seem to be some sort of special IDs. Is there any other code in win32k (or in your already-implemented IME code) that use these? maybe there are some #defines for these?

Comment (HBelusca) [reply win32ss/user/ntuser/ime.c:102]: Seems like 0 == just current thread; > 0 == that thread; -1 == all the threads (+ process)

File: win32ss/user/ntuser/ime.c [hunk 2]
```diff
@@ -80,14 +80,72 @@ NtUserCheckImeHotKey(
...
+
+        for (pti = ppi->ptiList; pti; )
+        {
+            pti->TIF_flags |= TIF_DISABLEIME;
+
+            if (pti->spwndDefaultIme)
+            {
+                co_UserDestroyWindow(pti->spwndDefaultIme);
+                pti->spwndDefaultIme = NULL;
+                pti = ptiCurrent->ppi->ptiList;
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:115]: looks like it's equivalent to:
```suggestion
                pti = ppi->ptiList;
```
(is it true that neither `ptiCurrent` nor `ppi` seems to change here? or not?)
But if this is so, then does this mean you restart looping from the beginning of the list?
If any case this definitively need a comment.

File: win32ss/user/ntuser/ime.c [hunk 3]
```diff
@@ -80,14 +80,72 @@ NtUserCheckImeHotKey(
...
+    else
+    {
+        if (dwThreadID == 0)
+        {
+            pti = ptiCurrent;
+        }
+        else
+        {
+            pti = IntTID2PTI(UlongToHandle(dwThreadID));
+            if (!pti || pti->ppi != ptiCurrent->ppi)
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:131]: Add a comment that the thread needs to reside in the current process.

File: win32ss/user/ntuser/ime.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c]: Simplified.

File: win32ss/user/ntuser/ime.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:133]: Added.

File: win32ss/user/ntuser/ime.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:102]: > Is there any other code in win32k

Nope.

Comment (katahiromz) [reply win32ss/user/ntuser/ime.c]: Defined `INVALID_THREAD_ID` and used it.

File: win32ss/user/ntuser/ime.c [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include <win32k.h>
 DBG_DEFAULT_CHANNEL(UserMisc);
 
+#define INVALID_THREAD_ID  0xFFFFFFFF
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:13]:
```suggestion
#define INVALID_THREAD_ID  ((ULONG)-1)
```

File: win32ss/user/ntuser/ime.c [hunk 5]
```diff
@@ -80,14 +81,74 @@ NtUserCheckImeHotKey(
...
+
+        for (pti = ppi->ptiList; pti; )
+        {
+            pti->TIF_flags |= TIF_DISABLEIME;
+
+            if (pti->spwndDefaultIme)
+            {
+                co_UserDestroyWindow(pti->spwndDefaultIme);
+                pti->spwndDefaultIme = NULL;
+                pti = ppi->ptiList;
```
Comment (HBelusca) [inline win32ss/user/ntuser/ime.c:116]: So, why do we need to restart looping through all the list?

File: win32ss/user/ntuser/ntuser.h [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #define RETURN(value) { _ret_ = value; goto _cleanup_; }
 #define CLEANUP /*unreachable*/ ASSERT(FALSE); _cleanup_
 #define END_CLEANUP return _ret_;
-
+#define IS_IMM_MODE() (gpsi && (gpsi->dwSRVIFlags & SRVINFO_IMM32))
```
Comment (HBelusca) [inline win32ss/user/ntuser/ntuser.h:7]: Shouldn't this macro need to go into an imm specific header?

File: win32ss/user/ntuser/ime.c (hunk 5 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:116]: I don't know.

File: win32ss/user/ntuser/ntuser.h (hunk 6 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ntuser.h:7]: This macro is specific to `win32k`.

File: win32ss/user/ntuser/ime.c (hunk 5 above)
Comment (katahiromz) [reply win32ss/user/ntuser/ime.c:116]: `ppi->ptiList` may be changed.

Comment (ThFabba) [reply win32ss/user/ntuser/ime.c]: `co_` functions call out to user mode, meaning the user lock will be released and threads could have been added or removed.

**Pull request merged**